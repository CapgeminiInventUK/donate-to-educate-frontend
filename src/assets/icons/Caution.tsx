import type { FC } from 'react';

const Caution: FC = () => (
  <svg width="23" height="23" viewBox="0 0 23 23" fill="none" xmlns="http://www.w3.org/2000/svg">
    <path
      d="M11.5 11.5V6.5M11.5 15.2061V15.25M21.5 11.5C21.5 17.0228 17.0228 21.5 11.5 21.5C5.97715 21.5 1.5 17.0228 1.5 11.5C1.5 5.97715 5.97715 1.5 11.5 1.5C17.0228 1.5 21.5 5.97715 21.5 11.5Z"
      stroke="#FEFCFD"
      strokeWidth="2"
      strokeLinecap="round"
      strokeLinejoin="round"
    />
  </svg>
);

export default Caution;
