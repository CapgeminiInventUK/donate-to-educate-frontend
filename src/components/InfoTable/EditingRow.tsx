import { FC, useState } from 'react';
import TextInput from '../TextInput/TextInput';
import FormButton from '../FormButton/FormButton';
import styles from './InfoTable.module.scss';
import { EditingRowProps } from '@/types/props';
import { phoneNumberRegex } from '@/utils/globals';
import { phone as validatePhoneNumber } from 'phone';

const EditingRow: FC<EditingRowProps> = ({
  field,
  value,
  setNewTableValues,
  setEditingKey,
  originalTableValues,
  onChange,
  refetch,
}) => {
  const [errorMessage, setErrorMessage] = useState<string>();
  const [editingValue, setEditingValue] = useState<string>();

  const onCancel = (): void => {
    setNewTableValues(originalTableValues);
    setErrorMessage && setErrorMessage(undefined);
    setEditingKey(undefined);
  };

  const onSave = (field: string): void => {
    if (
      field.toLowerCase() === 'phone' &&
      (!phoneNumberRegex.test(String(editingValue)) ||
        !validatePhoneNumber(String(editingValue), { country: 'GBR', validateMobilePrefix: false })
          .isValid)
    ) {
      setErrorMessage('Invalid phone number');
      return;
    }
    refetch &&
      void refetch().then(() => {
        setEditingKey(undefined);
        setNewTableValues(originalTableValues);
        setEditingValue(undefined);
      });
  };
  return (
    <div className={styles.editContainer}>
      <TextInput
        ariaLabel={`${field}-edit`}
        value={editingValue ?? value}
        onChange={(value) => {
          setEditingValue(value);
          onChange && onChange(field, value);
        }}
        errorMessage={errorMessage}
        className={styles.noMarginBottom}
      />
      <div className={`${styles.editButtons} ${errorMessage ? styles.marginTop : ''}`}>
        <FormButton
          className={styles.formButton}
          text="Save"
          theme={'formButtonGreen'}
          onClick={() => void onSave(field)}
          ariaLabel="save"
        />
        <FormButton
          className={styles.formButton}
          text="Cancel"
          theme="formButtonGrey"
          onClick={onCancel}
          ariaLabel="cancel"
        />
      </div>
    </div>
  );
};
export default EditingRow;
