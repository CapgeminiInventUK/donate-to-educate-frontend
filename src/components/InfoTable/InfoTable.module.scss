@use '@assets/global.scss' as g;

.infoTable {
  display: flex;
  flex-direction: column;
  background: g.$dark-blue;
  padding: 0 3%;
  border-radius: 10px;

  h3 {
    margin: 40px 0 20px 5px;
    display: flex;
    align-items: center;
    gap: 20px;

    .icon {
      display: flex;
      align-items: center;
    }

    .title {
      display: flex;
      align-items: center;
    }
  }

  .row {
    display: grid;
    grid-template-columns: 1fr 2fr 1fr;
    align-items: center;
    padding: 20px 0;
    border-bottom: 1px solid g.$primary-dark-blue;

    &:first-child {
      padding-top: 50px;
    }

    &:last-child {
      margin-bottom: 40px;
    }

    .key {
      font-weight: bold;
    }

    .value {
      text-align: left;

      .pill {
        display: inline-block;
        padding: 4px 20px;
        background-color: g.$pill-green-background;
        color: g.$pill-text-white;
        border-radius: 20px;
        font-weight: bold;
        text-align: center;
      }
    }

    .edit {
      text-align: center;
      font-size: 16px;
    }

    .delete {
      border-radius: 2px;
      border: 1px solid g.$border-black;
      padding: 8px 0;
      font-weight: 700;
      justify-self: end;
      color: g.$delete-red;
      background: g.$delete-grey;
      box-shadow: 2px 2px 0 2px g.$greyscale;
      min-width: 116px;
      height: 45px;
      cursor: pointer;
      font-size: 18px;

      &:hover {
        filter: brightness(0.9);
      }
    }
  }
}
