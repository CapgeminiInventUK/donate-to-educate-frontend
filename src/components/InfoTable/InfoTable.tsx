import React from 'react';
import styles from './InfoTable.module.scss';

interface InfoTableProps {
  tableValues: Record<string, string>;
  editableKeys?: string[];
  isAccounts?: boolean;
  isDelete?: boolean;
  title?: string;
  icon?: JSX.Element;
  className?: string;
  rowClassName?: string;
}

const InfoTable: React.FC<InfoTableProps> = ({
  tableValues,
  editableKeys = [],
  isAccounts = false,
  isDelete = false,
  title,
  icon,
  className,
  rowClassName,
}) => {
  return (
    <div className={`${styles.infoTable} ${className ? className : ''}`}>
      {(title ?? icon) && (
        <h3>
          <span className={styles.icon}>{icon}</span>
          <span className={styles.title}>{title}</span>
        </h3>
      )}
      {Object.entries(tableValues).map(([key, value]) => (
        <div className={`${styles.row} ${rowClassName ? rowClassName : ''}`} key={key}>
          <div className={styles.key}>{key}</div>
          <div className={styles.value}>
            {value ? value : isAccounts && <span className={styles.pill}>FREE</span>}
          </div>
          {editableKeys.includes(key) && (
            <a href="#" className={styles.edit}>
              Edit
            </a>
          )}
          {isDelete && (
            <span
              className={styles.delete}
              onClick={() => window.alert('This function is in development')}
            >
              Delete
            </span>
          )}
        </div>
      ))}
    </div>
  );
};

export default InfoTable;
