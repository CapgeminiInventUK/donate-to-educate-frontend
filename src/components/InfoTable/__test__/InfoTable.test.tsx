import { render } from '@testing-library/react';
import InfoTable from '../InfoTable';
import Caution from '../../../assets/icons/Caution';
import userEvent from '@testing-library/user-event';

describe('Info table', () => {
  it('should render component', () => {
    const props = {
      tableValues: {
        Name: 'Name',
        Email: 'test@test.com',
        'Job title or role': 'Job',
        Phone: '07777777777',
        Department: 'Department',
      },
      editableKeys: ['Job title or role', 'Department', 'Phone'],
      isAccounts: false,
      isDelete: false,
    };
    expect(render(<InfoTable {...props} />)).toMatchSnapshot();
  });

  it('should render component with title and icon when present', () => {
    const props = {
      tableValues: {
        'Local authority': 'LA',
        'Your account': 'test@test.com',
      },
      editableKeys: [],
      isAccounts: false,
      isDelete: true,
      title: 'Danger zone',
      className: 'deleteTable',
      rowClassName: 'deleteTableRow',
      icon: <Caution />,
    };
    expect(render(<InfoTable {...props} />)).toMatchSnapshot();
  });

  it('should render component with user accounts', () => {
    const props = {
      tableValues: {
        'Account 1': 'Account user one',
        'Account 2': 'Account user two',
        'Account 3': '',
      },
      editableKeys: [],
      isAccounts: true,
      isDelete: false,
    };
    expect(render(<InfoTable {...props} />)).toMatchSnapshot();
  });

  it('should handle Edit', async () => {
    const onEdit = vi.fn();
    const props = {
      tableValues: {
        Name: 'Name',
        Email: 'test@test.com',
        'Job title or role': 'Job',
        Phone: '07777777777',
        Department: 'Department',
      },
      editableKeys: ['Job title or role', 'Department', 'Phone'],
      isAccounts: false,
      isDelete: false,
      onEdit,
    };
    const { getAllByRole } = render(<InfoTable {...props} />);

    const editButton = getAllByRole('button', { name: 'edit-button' })[0];

    await userEvent.click(editButton);
    expect(onEdit).toHaveBeenCalled();
  });

  it('should handle Delete', async () => {
    const onDelete = vi.fn();
    const props = {
      tableValues: {
        'Local authority': 'LA',
        'Your account': 'test@test.com',
      },
      editableKeys: [],
      isAccounts: false,
      isDelete: true,
      title: 'Danger zone',
      className: 'deleteTable',
      rowClassName: 'deleteTableRow',
      icon: <Caution />,
      onDelete,
    };
    const { getAllByRole } = render(<InfoTable {...props} />);

    const editButton = getAllByRole('button', { name: 'delete-button' })[0];

    await userEvent.click(editButton);
    expect(onDelete).toHaveBeenCalled();
  });
});
