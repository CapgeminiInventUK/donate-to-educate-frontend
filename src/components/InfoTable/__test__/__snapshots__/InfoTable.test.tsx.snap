// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Info table > should render component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_infoTable_f1f6df "
      >
        <div
          class="_row_f1f6df "
        >
          <div
            class="_key_f1f6df"
          >
            Name
          </div>
          <div
            class="_value_f1f6df"
          >
            Name
          </div>
        </div>
        <div
          class="_row_f1f6df "
        >
          <div
            class="_key_f1f6df"
          >
            Email
          </div>
          <div
            class="_value_f1f6df"
          >
            test@test.com
          </div>
        </div>
        <div
          class="_row_f1f6df "
        >
          <div
            class="_key_f1f6df"
          >
            Job title or role
          </div>
          <div
            class="_value_f1f6df"
          >
            Job
          </div>
          <button
            aria-label="edit-button"
            class="_link_f8f296 _edit_f1f6df "
          >
            Edit
          </button>
        </div>
        <div
          class="_row_f1f6df "
        >
          <div
            class="_key_f1f6df"
          >
            Phone
          </div>
          <div
            class="_value_f1f6df"
          >
            07777777777
          </div>
          <button
            aria-label="edit-button"
            class="_link_f8f296 _edit_f1f6df "
          >
            Edit
          </button>
        </div>
        <div
          class="_row_f1f6df "
        >
          <div
            class="_key_f1f6df"
          >
            Department
          </div>
          <div
            class="_value_f1f6df"
          >
            Department
          </div>
          <button
            aria-label="edit-button"
            class="_link_f8f296 _edit_f1f6df "
          >
            Edit
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_infoTable_f1f6df "
    >
      <div
        class="_row_f1f6df "
      >
        <div
          class="_key_f1f6df"
        >
          Name
        </div>
        <div
          class="_value_f1f6df"
        >
          Name
        </div>
      </div>
      <div
        class="_row_f1f6df "
      >
        <div
          class="_key_f1f6df"
        >
          Email
        </div>
        <div
          class="_value_f1f6df"
        >
          test@test.com
        </div>
      </div>
      <div
        class="_row_f1f6df "
      >
        <div
          class="_key_f1f6df"
        >
          Job title or role
        </div>
        <div
          class="_value_f1f6df"
        >
          Job
        </div>
        <button
          aria-label="edit-button"
          class="_link_f8f296 _edit_f1f6df "
        >
          Edit
        </button>
      </div>
      <div
        class="_row_f1f6df "
      >
        <div
          class="_key_f1f6df"
        >
          Phone
        </div>
        <div
          class="_value_f1f6df"
        >
          07777777777
        </div>
        <button
          aria-label="edit-button"
          class="_link_f8f296 _edit_f1f6df "
        >
          Edit
        </button>
      </div>
      <div
        class="_row_f1f6df "
      >
        <div
          class="_key_f1f6df"
        >
          Department
        </div>
        <div
          class="_value_f1f6df"
        >
          Department
        </div>
        <button
          aria-label="edit-button"
          class="_link_f8f296 _edit_f1f6df "
        >
          Edit
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Info table > should render component with title and icon when present 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_infoTable_f1f6df deleteTable"
      >
        <h3>
          <span
            class="_icon_f1f6df"
          >
            <svg
              fill="none"
              height="23"
              viewBox="0 0 23 23"
              width="23"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M11.5 11.5V6.5M11.5 15.2061V15.25M21.5 11.5C21.5 17.0228 17.0228 21.5 11.5 21.5C5.97715 21.5 1.5 17.0228 1.5 11.5C1.5 5.97715 5.97715 1.5 11.5 1.5C17.0228 1.5 21.5 5.97715 21.5 11.5Z"
                stroke="#FEFCFD"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
          </span>
          <span
            class="_title_f1f6df"
          >
            Danger zone
          </span>
        </h3>
        <div
          class="_row_f1f6df deleteTableRow"
        >
          <div
            class="_key_f1f6df"
          >
            Local authority
          </div>
          <div
            class="_value_f1f6df"
          >
            LA
          </div>
          <button
            aria-label="delete-button"
            class="_light_f8f296 _delete_f1f6df "
          >
            Delete
          </button>
        </div>
        <div
          class="_row_f1f6df deleteTableRow"
        >
          <div
            class="_key_f1f6df"
          >
            Your account
          </div>
          <div
            class="_value_f1f6df"
          >
            test@test.com
          </div>
          <button
            aria-label="delete-button"
            class="_light_f8f296 _delete_f1f6df "
          >
            Delete
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_infoTable_f1f6df deleteTable"
    >
      <h3>
        <span
          class="_icon_f1f6df"
        >
          <svg
            fill="none"
            height="23"
            viewBox="0 0 23 23"
            width="23"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11.5 11.5V6.5M11.5 15.2061V15.25M21.5 11.5C21.5 17.0228 17.0228 21.5 11.5 21.5C5.97715 21.5 1.5 17.0228 1.5 11.5C1.5 5.97715 5.97715 1.5 11.5 1.5C17.0228 1.5 21.5 5.97715 21.5 11.5Z"
              stroke="#FEFCFD"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </span>
        <span
          class="_title_f1f6df"
        >
          Danger zone
        </span>
      </h3>
      <div
        class="_row_f1f6df deleteTableRow"
      >
        <div
          class="_key_f1f6df"
        >
          Local authority
        </div>
        <div
          class="_value_f1f6df"
        >
          LA
        </div>
        <button
          aria-label="delete-button"
          class="_light_f8f296 _delete_f1f6df "
        >
          Delete
        </button>
      </div>
      <div
        class="_row_f1f6df deleteTableRow"
      >
        <div
          class="_key_f1f6df"
        >
          Your account
        </div>
        <div
          class="_value_f1f6df"
        >
          test@test.com
        </div>
        <button
          aria-label="delete-button"
          class="_light_f8f296 _delete_f1f6df "
        >
          Delete
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Info table > should render component with user accounts 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_infoTable_f1f6df "
      >
        <div
          class="_row_f1f6df "
        >
          <div
            class="_key_f1f6df"
          >
            Account 1
          </div>
          <div
            class="_value_f1f6df"
          >
            Account user one
          </div>
        </div>
        <div
          class="_row_f1f6df "
        >
          <div
            class="_key_f1f6df"
          >
            Account 2
          </div>
          <div
            class="_value_f1f6df"
          >
            Account user two
          </div>
        </div>
        <div
          class="_row_f1f6df "
        >
          <div
            class="_key_f1f6df"
          >
            Account 3
          </div>
          <div
            class="_value_f1f6df"
          >
            <span
              class="_pill_f1f6df"
            >
              FREE
            </span>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_infoTable_f1f6df "
    >
      <div
        class="_row_f1f6df "
      >
        <div
          class="_key_f1f6df"
        >
          Account 1
        </div>
        <div
          class="_value_f1f6df"
        >
          Account user one
        </div>
      </div>
      <div
        class="_row_f1f6df "
      >
        <div
          class="_key_f1f6df"
        >
          Account 2
        </div>
        <div
          class="_value_f1f6df"
        >
          Account user two
        </div>
      </div>
      <div
        class="_row_f1f6df "
      >
        <div
          class="_key_f1f6df"
        >
          Account 3
        </div>
        <div
          class="_value_f1f6df"
        >
          <span
            class="_pill_f1f6df"
          >
            FREE
          </span>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
